miniapp/miniapp_tsgemm.cpp

```diff
@@ -0,0 +1,484 @@
+      if (rank == 0) {
+        std::cout << cfin_sum << '\n';
+      }
+    }
+  }
+
+  // Upon exit the mpi/user polling RAII object will stop polling
+  return hpx::finalize();
+}
+
+// Example usage:
+//
+//   mpirun -np 1 tsgemm --len_m      100  --len_n      100  --len_k  10000
+//                       --tile_m      32  --tile_n      32
+//                       --pgrid_rows   1  --pgrid_cols   1
+//
+int main(int argc, char** argv) {
+  // Flush printf
+  setbuf(stdout, nullptr);
+
+  // Initialize MPI
+  int thd_required = MPI_THREAD_MULTIPLE;
+  int thd_provided;
+  MPI_Init_thread(&argc, &argv, thd_required, &thd_provided);
+
+  if (thd_required != thd_provided) {
+    std::fprintf(stderr, "Provided MPI threading model does not match the required one.\n");
+    MPI_Abort(MPI_COMM_WORLD, 1);
+  }
+
+  // Declare options before creating resource partitioner
+  namespace po = hpx::program_options;
+  po::options_description desc_cmdline = ::init_desc();
+  po::variables_map vm;
+  po::store(po::command_line_parser(argc, argv).allow_unregistered().options(desc_cmdline).run(), vm);
+  std::string setup = vm["setup"].as<std::string>();
+
+  if (setup == "mpi_pool") {
+    int num_procs;
+    MPI_Comm_size(MPI_COMM_WORLD, &num_procs);
+    if (num_procs > 1) {  // if more than 1 process
+      // Create a thread pool for MPI work and add (enabled) PUs on the first core
+      //
+      // Note: Thread pools must be declared before starting the runtime
+      hpx::resource::partitioner rp(desc_cmdline, argc, argv);
+      if (rp.numa_domains()[0].cores().size() != 1) {  // if more than 1 core
+        rp.create_thread_pool("mpi");
+        rp.add_resource(rp.numa_domains()[0].cores()[0].pus(), "mpi");
+      }
+    }
+  }
+
+  // Start the HPX runtime
+  hpx::init(desc_cmdline, argc, argv);
+
+  MPI_Finalize();
+}
+
+namespace {
+
+// Send Cini tile to the process it belongs (`tile_rank`)
+void send_tile(ConstTileType const& c_tile, SizeType tile_rank, SizeType tag, MPI_Comm comm) {
+  SizeType num_elements = c_tile.ld() * c_tile.size().cols();
+  void const* buf = c_tile.ptr(TileElementIndex(0, 0));
+  MPI_Datatype dtype = dlaf::comm::mpi_datatype<ScalarType>::type;
+
+  MPI_Request req;
+  MPI_Isend(buf, num_elements, dtype, tile_rank, tag, comm, &req);
+  hpx::util::yield_while([&req] {
+    int flag;
+    MPI_Test(&req, &flag, MPI_STATUS_IGNORE);
+    return flag == 0;
+  });
+}
+
+// Reduce into Cini tile from all other processes
+void recv_tile(TileType& c_tile, int this_rank, SizeType tag, MPI_Comm comm) {
+  int nprocs = Communicator(MPI_COMM_WORLD).size() - 1;
+  SizeType nelems = c_tile.ld() * c_tile.size().cols();
+  ScalarType* tile_buf = c_tile.ptr(TileElementIndex(0, 0));
+  MPI_Datatype dtype = dlaf::comm::mpi_datatype<ScalarType>::type;
+
+  // Note: these allocations should be better made from a pool
+  std::vector<MPI_Request> reqs(nprocs);
+  std::vector<ScalarType> staging_buf(nprocs * nelems);
```

Comment:
I looked at `dlaf::common::Buffer` and it is indeed a better fit for my us case since it utilizes `std::make_unique<T[]>()`. However, in the future, the allocations should ideally be made from a pool. `std::make_unique<T[]>(N)` uses `malloc()` which is often expensive.

Reply:
Good point, when I designed `Buffer` I thought about it but I left the implementation for the future (when the Pool from #96 will be available).

Just for reference, you can give a look to the last parts of this messages, where we quickly discussed about it

https://github.com/eth-cscs/DLA-Future/pull/155#issuecomment-595769394
https://github.com/eth-cscs/DLA-Future/pull/155#discussion_r392228763